oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/RefreshTokenOAuth2AuthorizedClientProvider.java

```diff
@@ -0,0 +1,141 @@
+		this.clientRegistrationRepository = clientRegistrationRepository;
+		this.authorizedClientRepository = authorizedClientRepository;
+	}
+
+	/**
+	 * Attempt to re-authorize the {@link OAuth2AuthorizationContext#getClientRegistration() client} in the provided {@code context}.
+	 * Returns {@code null} if re-authorization is not supported,
+	 * e.g. the {@link OAuth2AuthorizedClient#getRefreshToken() refresh token} is not available for the
+	 * {@link OAuth2AuthorizationContext#getAuthorizedClient() authorized client}.
+	 *
+	 * <p>
+	 * The following {@link OAuth2AuthorizationContext#getAttributes() context attributes} are supported:
+	 * <ol>
+	 *  <li>{@code "javax.servlet.http.HttpServletRequest"} (required) - the {@code HttpServletRequest}</li>
+	 *  <li>{@code "javax.servlet.http.HttpServletResponse"} (required) - the {@code HttpServletResponse}</li>
+	 *  <li>{@code "org.springframework.security.oauth2.client.REQUEST_SCOPE"} (optional) - a space-delimited or comma-delimited {@code String} of scope(s) to be requested by the {@link OAuth2AuthorizationContext#getClientRegistration() client}</li>
+	 * </ol>
+	 *
+	 * @param context the context that holds authorization-specific state for the client
+	 * @return the {@link OAuth2AuthorizedClient} or {@code null} if re-authorization is not supported
+	 */
+	@Override
+	@Nullable
+	public OAuth2AuthorizedClient authorize(OAuth2AuthorizationContext context) {
+		Assert.notNull(context, "context cannot be null");
+		if (!context.reauthorizationRequired() || context.getAuthorizedClient().getRefreshToken() == null) {
+			return null;
+		}
+
+		HttpServletRequest request = context.getAttribute(HTTP_SERVLET_REQUEST_ATTRIBUTE_NAME);
+		HttpServletResponse response = context.getAttribute(HTTP_SERVLET_RESPONSE_ATTRIBUTE_NAME);
+		Assert.notNull(request, "The context attribute cannot be null '" + HTTP_SERVLET_REQUEST_ATTRIBUTE_NAME + "'");
+		Assert.notNull(response, "The context attribute cannot be null '" + HTTP_SERVLET_RESPONSE_ATTRIBUTE_NAME + "'");
+
+		String requestScope = context.getAttribute(REQUEST_SCOPE_ATTRIBUTE_NAME);
+		Set<String> scopes = null;
+		if (!StringUtils.isEmpty(requestScope)) {
+			String delimiter = requestScope.indexOf(',') != -1 ? "," : " ";
+			scopes = Arrays.stream(StringUtils.delimitedListToStringArray(requestScope, delimiter, " ")).collect(Collectors.toSet());
+		}
+
+		OAuth2RefreshTokenGrantRequest refreshTokenGrantRequest =
+				new OAuth2RefreshTokenGrantRequest(context.getAuthorizedClient(), scopes);
+		OAuth2AccessTokenResponse tokenResponse =
+				this.accessTokenResponseClient.getTokenResponse(refreshTokenGrantRequest);
+
+		OAuth2AuthorizedClient authorizedClient = new OAuth2AuthorizedClient(
+				context.getClientRegistration(),
+				context.getPrincipal().getName(),
+				tokenResponse.getAccessToken(),
+				tokenResponse.getRefreshToken());
+
+		this.authorizedClientRepository.saveAuthorizedClient(
```

Comment:
It is quite surprising to me that the provider will save the result but not try and load it first. I think the API should not save the authorization. I am even more convinced the API should not do save without trying to perform a lookup as I find this asymmetry confusing.

Reply:
> It is quite surprising to me that the provider will save the result but not try and load it first

The `OAuth2AuthorizedClient` is supplied via the context - why would it look it up again? I'm not sure I follow?

Reply:
That is what I'm trying to get at. It is strange to require the user to lookup the inputs via the repository API and the provider will save the result.

I think we should either have the API allow looking up the OAuth2AuthorizedClient or (likely preferred) not save the result.